Feature/backend tests for /api/posts/{postId} endpoint and dbHelper

galeza opened the pull request:
### What does this pull request aim to achieve? 💯

_Please be concise and link any related issue(s):_

Add tests for /api/posts/{postId} and dbHelper

- [x] I have checked that no one else is working on similar changes.
- [x] I have read the latest [**README**](README.md) and understand the development workflow.
- [x] The name of this branch is: **`feature/<branch_name>`**, or **`hotfix/<branch_name>`** if this is a hotfix, where `<branch_name>` briefly describes the issue(s) resolved and is prefixed with the issue number(s) (e.g. `1127-1130-update-git-branching-model`). The branch is created in a cloned version of this repo, **not a forked repo**.
- [x] This branch is rebased/merged with the **latest staging** (or the **latest production** if it is a hotfix).
- [x] There are no merge conflicts.
- [x] There are no console warnings and errors.
- [x] On the right hand side of this PR, I have linked the appropriate issue(s) under "Issues" and added a project progress under "Projects".
- [x] The title describes the issue(s) being addressed using format: **`<issue_numbers> - <brief_issue_description>`** (e.g. `#1127/#1130 - Update Git Branching Model`)

galeza committed:
Add mongoose

galeza committed:
Add dbHelper and models

galeza committed:
Start adding test for get post endpoint

galeza committed:
Add test user gets a post by ID

galeza committed:
Correct API endpoint for posts

galeza committed:
Add new methods to dbHelper, test with random ObjectID

galeza committed:
Remove comments from Post model

galeza commented:
Hi @jessica-lau 

If you have time, please look at this solution. Once you asked me about adding integration with DB for our tests.
I added dbHelper.js and I'm retrieving data from DB.
What you think about having more integration tests?

jessica-lau commented:
> Hi @jessica-lau
> 
> If you have time, please look at this solution. Once you asked me about adding integration with DB for our tests.
> I added dbHelper.js and I'm retrieving data from DB.
> What you think about having more integration tests?

hey @galeza I looked at the solution and I think the post.test.js should include code to set up a dataset in the local mongoDB server before the test cases so that they can run off on the local data. If we're using the live server, the tests can come out flaky since data will always be changing. What do you think about this? Yea, I agree it's necessary to have more integration tests.

Another question I have is how come a post.js file was created in the the mocha>models folder instead of importing the schema in Post.js from the backend directly like this:

I'm just curious since this can reduce the need to create additional files.

In post.test.js:

`const { model } = require("../backend/lib/models/Post")`

In docker-compose.yml:
```volumes:
      - ./mocha/test/:/app/test/
      - ./mocha/models/:/app/models/
      - ./mocha/utils/:/app/utils/
      - ./backend/lib/models/:/app/backend/lib/models

galeza committed:
Add insert method to the dbHelper

galeza committed:
Start modifing post tests - add a doc to the DB

galeza committed:
Correct beforeEach method

galeza committed:
Insert post into db and get result

galeza committed:
Add results objects comparison

galeza committed:
Remove document from posts collection

galeza committed:
Clean console logs

galeza committed:
Add Describe

galeza committed:
Correct validators

galeza committed:
Add test

galeza committed:
Remove models

galeza commented:
Hi,
I added dbHelper.js.
The flow is as follow:
Before - we connect to the DB
Before each - Insert a new document to the DB
Test - compare objects
After each - Delete a document
After - Close DB

When I compare object returned from DB and object returned by API I omit some values, API object contains additional data. But "author.id", "createdAt", "updatedAt", "_id" have different types. Let me know if you know how to bypass that. I was using JSON.stringify, but it doesn't work.

Two more tests are verifying endpoint with different values (random ObjectID and Int). Issue [API] - endpoint GET /api/posts/{postId} #2088 was created.

FYI. 
1. Two Login Tests fail + one cypress Login test. I think that they should be corrected in a different PR
2. Another thought - we need refactor validators.js. Maybe we can make it generic. Pass map with key value and compare them. If you agree I add a new task.

TimeaKovacs86 commented:
I got the same result running the tests:  Two Login Tests fail + one cypress Login test as you said @galeza 
But somehow I couldn't run mocha integration: keep getting the same error message : Error: Cannot find module 'lodash', do you have any idea why is that? What did I do wrong?

galeza commented:
> I got the same result running the tests: Two Login Tests fail + one cypress Login test as you said @galeza
> But somehow I couldn't run mocha integration: keep getting the same error message : Error: Cannot find module 'lodash', do you have any idea why is that? What did I do wrong?

@TimeaKovacs86 
I think that docker-compose up --build is necessary as I changed package and package-lock

Yes that was the issue! Thank you! Ran the tests again, got the same result as you and @jessica-lau  also.  Login test fails which, I agree, should be handled by a separate PR. Other than this... I don't know how you can bypass datatype differences. Sorry. :( Seems like I learn more than I can help...

jessica-lau commented on mocha/utils/apiEndpoints.js:
```diff
@@ -3,3 +3,6 @@ exports.feedbackEndpoint = "/api/feedback";
 exports.organisationsEndpoint = "/api/organisations";
 exports.postsEndpoint = "/api/posts";
+exports.postApiEndpoint = "/api/posts/{postId}";
+exports.getPostApiEndpoint = "/api/posts/";
```
Is this a duplicate? If I'm understanding correctly, I think line 8 "/api/posts" is being used in the test but not line 7 "/api/posts/".

galeza replied:
Corrected

jessica-lau commented on mocha/utils/validators.js:
```diff
@@ -39,3 +39,22 @@ exports.validateStatusCodeResponseBody = function (
+  expect(response).to.have.a.property("statusCode").to.be.equal(statusCode);
+  expect(response).to.have.a.property("status").to.be.equal(status);
+  expect(response).to.have.a.property("ok").to.be.equal(true);
+};
```
I think checking either the statusCode or status in the response is enough since both of them are the same but just stated differently, so no need for both. 200 is httpStatus.OK

galeza replied:
Corrected.

jessica-lau commented on mocha/utils/validators.js:
```diff
@@ -39,3 +39,22 @@ exports.validateStatusCodeResponseBody = function (
+  initial,
+  initialPropToOmit,
+  response,
+  responsPropToOmit,
```
Just a minor typo "responsPropToOmit" should be "responsePropToOmit"

galeza replied:
Corrected.

jessica-lau commented on mocha/utils/validators.js:
```diff
@@ -39,3 +39,22 @@ exports.validateStatusCodeResponseBody = function (
+) {
+  var result = _.isEqual(
+    _.omit(initial, initialPropToOmit),
+    _.omit(response, responsPropToOmit),
```
Just a minor typo "responsPropToOmit" should be "responsePropToOmit"

galeza replied:
Corrected.

jessica-lau commented on mocha/test/post.test.js:
```diff
@@ -0,0 +1,124 @@
+        APP_URL,
+        getPostApiEndpoint + "/" + postID,
+      );
+      validator.validateStatusCodeAndMessage(response, httpStatus.OK, 200);
```
I think checking either the statusCode or status in the response is enough since both of them are the same but just stated differently, so no need for both. 200 is httpStatus.OK

galeza replied:
Corrected.

jessica-lau commented on mocha/test/post.test.js:
```diff
@@ -0,0 +1,124 @@
+  visibility: "worldwide",
+};
+describe("GET /api/posts/{postId} endpoint - for a user that is NOT signed in", function () {
+  describe("GET /api/posts/{postId} endpoint - record is added to the DB and then retrived using API", function () {
```
"retrieved"

galeza replied:
Corrected.

jessica-lau commented on mocha/utils/apiEndpoints.js:
```diff
@@ -3,3 +3,6 @@ exports.feedbackEndpoint = "/api/feedback";
 exports.signUpApiEndpoint = "/api/auth/signup";
 exports.organisationsEndpoint = "/api/organisations";
 exports.postsEndpoint = "/api/posts";
+exports.postApiEndpoint = "/api/posts/{postId}";
```
Don't think "/api/posts/{postId}" endpoint is being used. Remove?

galeza replied:
Removed

jessica-lau reviewed (changes requested):
This looks great @galeza . I only got the login test fails which, I agree, should be handled by a separate PR. Also, your suggestion on refactoring the validators.js where applicable makes sense to simplify things. 

I don't know a way to bypass the different datatype for ["author.location", "author.id", "createdAt", "updatedAt", "_id"] but I can look into this when I have some time and see if I can find some sort of solution. 

One other thing -  could we possibly rename the test file from "post.test.js" to something like "postId.test.js"? There's another test called "posts.test.js" that checks posts exists in the /api/posts endpoint and the names are so close and can be confused with each other. Thanks!

galeza committed:
Refactor PostId

galeza committed:
Remove unused var

galeza committed:
Add generic validateResponse method

galeza committed:
Refactor existing tests

galeza committed:
Improve validators. Remove method

galeza commented:
@jessica-lau @TimeaKovacs86 
@jessica-lau I implemented your suggestions, please check.
I added also a generic validator that takes dict and validate kay-value. I removed all other methods. Please see and let me know if I miss anything.
All tests should pass.

JayashreeSamantasinghar reviewed (commented):
@galeza 
All mocha integration tests passed
<img width="960" alt="2020-12-17 (1)" src="https://user-images.githubusercontent.com/68357352/102559935-e31ba580-4085-11eb-85aa-b7cfbdebf49b.png">

TimeaKovacs86 commented:
@galeza  LGTM! Mocha tests pass! And as we discussed before there is only one cypress test failing, and gonna be fixed in a future PR! Great job! 

![backend test post endpoint mocha integration tests pass](https://user-images.githubusercontent.com/46457342/102608204-6d651780-4121-11eb-93fd-c97cc2a091a4.PNG)

jessica-lau reviewed (dismissed):
LGTM @galeza 

![image](https://user-images.githubusercontent.com/56090896/102648983-74e2eb80-4136-11eb-9fe1-7ee5d5a49bbe.png)

JayashreeSamantasinghar commented:
@galeza 
I was trying other methods to compare both files without omitting some keys. 
In validator.js----> exports.validateInitialAndResponseObject = function (initial, response) {
  _.isEqual(initial, response);
};
In postId.test.js------>  validator.validateInitialAndResponseObject(
        insertResult[0],
        response.body.post
);
Can we try in this way?

galeza commented:
> @galeza
> I was trying other methods to compare both files without omitting some keys.
> In validator.js----> exports.validateInitialAndResponseObject = function (initial, response) {
> _.isEqual(initial, response);
> };
> In postId.test.js------> validator.validateInitialAndResponseObject(
> insertResult[0],
> response.body.post
> );
> Can we try in this way?

Hi @JayashreeSamantasinghar 
The problem is that we have more data in the response.body.post that's why it was necessary to omit them (we have additional data like status, counters).

adorapham commented:
LGTM, i get the same result 
![image](https://user-images.githubusercontent.com/70918430/103232107-6fcc2c00-48ee-11eb-91d9-66c2eca69667.png)
